client/finality-grandpa/rpc/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+    pub threshold_weight: u64,
+    pub prevotes: Prevotes,
+    pub precommits: Precommits,
+}
+
+impl RoundState {
+    pub fn from(
+        round: u64,
+        round_state: &voter::report::RoundState<AuthorityId>,
+        voters: &HashSet<AuthorityId>,
+    ) -> Self {
+        let prevotes = &round_state.prevote_ids;
+        let missing_prevotes = voters.difference(&prevotes).cloned().collect();
+
+        let precommits = &round_state.precommit_ids;
+        let missing_precommits = voters.difference(&precommits).cloned().collect();
+
+        Self {
+            round,
+            total_weight: round_state.total_weight,
+            threshold_weight: round_state.threshold_weight,
+            prevotes: Prevotes {
+                current_weight: round_state.prevote_current_weight,
+                missing: missing_prevotes,
+            },
+            precommits: Precommits {
+                current_weight: round_state.precommit_current_weight,
+                missing: missing_precommits,
+            },
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ReportedRoundStates {
+    pub set_id: u64,
+    pub best: RoundState,
+    pub background: Vec<RoundState>,
+}
+
+impl ReportedRoundStates {
+    pub fn from<Hash, Block>(
+        voter_state: &SharedVoterState,
+        authority_set: &SharedAuthoritySet<Hash, Block>,
+    ) -> Result<Self, Error>
+    where
+        Hash: Debug + Clone + Eq + Send + Sync + 'static,
+        Block: BlockNumberOps + Send + Sync + 'static,
+    {
+        let voter_state = voter_state
+            .read()
+            .as_ref()
+            .map(|vs| vs.voter_state())
+            .ok_or(Error::EndpointNotReady)?;
+
+        let current_voters: HashSet<AuthorityId> = authority_set
+            .current_authorities()
+			.voters()
+            .iter()
+            .map(|p| p.0.clone())
+            .collect();
+
+        let best = {
+            let (round, round_state) = voter_state.best_round;
+            RoundState::from(round, &round_state, &current_voters)
+        };
+
+        let background = voter_state
+            .background_rounds
+            .iter()
+            .map(|(round, round_state)| RoundState::from(*round, round_state, &current_voters))
+            .collect();
+
+        Ok(Self {
+            set_id: authority_set.set_id(),
+            best,
+            background,
+        })
+    }
+}
+
+impl<Hash, Block: Send + Sync> GrandpaApi for GrandpaRpcHandler<Hash, Block>
+where
+    Hash: Debug + Clone + Eq + Send + Sync + 'static,
+    Block: BlockNumberOps + Send + Sync + 'static,
+{
+    fn round_state(&self) -> FutureResult<ReportedRoundStates> {
+        let round_states =
+            ReportedRoundStates::from(&self.shared_voter_state, &self.shared_authority_set);
+        let future = async move { round_states }.boxed();
+        Box::new(future.map_err(jsonrpc_core::Error::from).compat())
+    }
+}
```

Comment:
Initially you had a test stub here and I think it was nice test to have.

Reply:
I second that.